provider/cmd/crd2pulumi/generate.go

```diff
@@ -0,0 +1,209 @@
+	Group string
+	// Versions represents a mapping from each version name in the
+	// `spec.versions` list to its corresponding `openAPIV3Schema` field in the
+	// CRD YAML
+	Versions map[string]map[string]interface{}
+}
+
+func NewCustomResourceGenerator(language, yamlPath, outputPath string) (CustomResourceGenerator, error) {
+	if language != DotNet && language != Go && language != NodeJS && language != Python {
+		return CustomResourceGenerator{}, errors.New("invalid language: " + language)
+	}
+
+	yamlFile, err := ioutil.ReadFile(yamlPath)
+	if err != nil {
+		return CustomResourceGenerator{}, fmt.Errorf("read file %s: %v", yamlPath, err)
+	}
+
+	crd, err := UnmarshalYaml(yamlFile)
+	if err != nil {
+		return CustomResourceGenerator{}, fmt.Errorf("unmarshal %s: %v", yamlPath, err)
+	}
+
+	underscoreFields(crd.Object)
+
+	apiVersion := crd.GetAPIVersion()
+	if apiVersion != v1beta1 && apiVersion != v1 {
+		return CustomResourceGenerator{}, fmt.Errorf("invalid apiVersion %s: %v", apiVersion, err)
+	}
+
+	versions := map[string]map[string]interface{}{}
+
+	validation, foundValidation, _ := unstruct.NestedMap(crd.Object, "spec", "validation", "openAPIV3Schema")
+	if foundValidation { // If present, use the top-level schema to validate all versions
+		versionMaps, _, _ := NestedMapSlice(crd.Object, "spec", "versions")
+		for _, version := range versionMaps {
+			name, _, _ := unstruct.NestedString(version, "name")
+			versions[name] = validation
+		}
+	} else { // Otherwise use per-version schemas to validate each version
+		versionMaps, _, _ := NestedMapSlice(crd.Object, "spec", "versions")
+		for _, version := range versionMaps {
+			name, _, _ := unstruct.NestedString(version, "name")
+			schema, _, _ := unstruct.NestedMap(version, "schema", "openAPIV3Schema")
+			versions[name] = schema
+		}
+	}
+
+	kind, _, _ := unstruct.NestedString(crd.Object, "spec", "names", "kind")
+	plural, _, _ := unstruct.NestedString(crd.Object, "spec", "names", "plural")
+	group, _, _ := unstruct.NestedString(crd.Object, "spec", "group")
+
+	if outputPath == "" {
+		defaultOutputPath := func(yamlPath, plural, language string) string {
+			var extension string
+			switch language {
+			case NodeJS:
+				extension = "ts"
+			case DotNet:
+				extension = "cs"
+			case Python:
+				extension = "py"
+			case Go:
+				extension = "go"
+			}
+			outputFileName := plural + "." + extension
+			return path.Join(filepath.Dir(yamlPath), outputFileName)
+		}
+		outputPath = defaultOutputPath(yamlPath, plural, language)
+	}
+
+	customResourceGenerator := CustomResourceGenerator{
+		CustomResourceDefinition: crd,
+		OutputPath:               outputPath,
+		Language:                 language,
+		APIVersion:               apiVersion,
+		Kind:                     kind,
+		Plural:                   plural,
+		Group:                    group,
+		Versions:                 versions,
+	}
+	return customResourceGenerator, nil
+}
+
+func (gen *CustomResourceGenerator) GetName() string {
```

Comment:
nit: In Go, this type of method usually doesn't include a `Get` prefix, so this would be `Name()` rather than `GetName()`